core/backend/src/CatalogIModel.ts

```diff
@@ -0,0 +1,327 @@
+    });
+    container.disconnect();
+    return cloudContainerProps;
+  }
+
+  /** Acquire the write lock for a CatalogIModel container. Only one person may obtain the write lock at a time.
+   * @note this requires "write" authorization to the container
+   */
+  export async function acquireWriteLock(args: {
+    /** The id of the container */
+    containerId: string,
+    /**
+     * The name of the individual acquiring the lock. This will be shown to others who attempt to acquire the lock while it is held.
+     * It is also stored in the "lastEditedBy" field of the manifest of any new version edited while the lock is held.
+     */
+    username: string;
+  }): Promise<void> {
+    const container = await getWriteableContainer(args.containerId);
+    return CloudSqlite.acquireWriteLock({ container, user: args.username });
+  }
+
+  /** Release the write lock on a CatalogIModel container. This uploads all changes made while the lock is held, so they become visible to other users. */
+  export async function releaseWriteLock(args: {
+    /** The id of the container */
+    containerId: string,
+    /** If true, abandon all local changes before releasing the lock */
+    abandon?: true
+  }): Promise<void> {
+    const container = await getWriteableContainer(args.containerId);
+    if (args.abandon)
+      container.abandonChanges();
+    CloudSqlite.releaseWriteLock(container);
+  }
+
+  /** Open a CatalogIModel for write access.
+   * @note Once a version of a CatalogIModel has been published (i.e. the write lock has been released), it is no longer editable, *unless* it is a prerelease version.
+   * @note the write lock must be held for this operation to succeed
+   */
+  export async function openEditable(args: CatalogIModelTypes.OpenArgs): Promise<StandaloneDb & EditCatalog> {
+    const dbName = catalogDbNameWithDefault(args.dbName);
+    if (undefined === args.containerId) // local file?
+      return EditableCatalogDb.openFile(dbName, OpenMode.ReadWrite, args) as EditableCatalogDb;
+
+    const container = await getWriteableContainer(args.containerId);
+    ensureLocked(container, "open a Catalog for editing"); // editing Catalogs requires the write lock
+
+    // look up the full name with version
+    const dbFullName = CloudSqlite.querySemverMatch({ container, dbName, version: args.version ?? "*" });
+    if (!CloudSqlite.isSemverEditable(dbFullName, container))
+      CloudSqliteError.throwError("already-published", { message: "Catalog has already been published and is not editable. Make a new version first.", ...args })
+
+    if (args.prefetch)
+      CloudSqlite.startCloudPrefetch(container, dbFullName);
+
+    return EditableCatalogDb.openFile(dbFullName, OpenMode.ReadWrite, { container, ...args }) as EditableCatalogDb;
+  }
+
+  /**
+   * Open a CatalogIModel for read access.
+   */
+  export async function openReadonly(args: CatalogIModelTypes.OpenArgs): Promise<StandaloneDb & ReadCatalog> {
```

Comment:
I'm curious about the motivation for exposing this as `StandaloneDb & ReadCatalog` vs just exposing `CatalogDb` as a subclass of `StandaloneDb`. Can somebody do something useful with a `ReadCatalog` by itself?

Reply:
I didn't want to export `CatalogDb` so nobody could create instances of it (or `EditableCatalogDb`) directly.

Reply:
If that's important on the back-end, why does the front-end just provide `CatalogConnection` as a subclass of `BriefcaseConnection`?
It looks like if I opened a `CatalogConnection` as read-only and called its `updateCatalogManifest` method, I would get a very confusing error from the backend (via `CatalogIModelHandler`) about `updateCatalogManifest` being `undefined`. (`findCatalogByKey` asserts the result is an `EditCatalog` but it's only a `CatalogDb` if opened read-only).

Reply:
See if you like [this](https://github.com/iTwin/itwinjs-core/pull/7933/commits/c60094e9dbf4c618d503ce1ae14de4a3d0db54cf) better for backend. `CatalogIModel` extends *only* the non-static *interface* of `StandaloneDb` and adds catalog-specific methods.

Reply:
Made similar changes on front-end.